packages/alchemy/src/react/hooks/useSignTypedData.ts

```diff
@@ -0,0 +1,70 @@
+"use client";
+
+import {
+  useMutation,
+  type UseMutateAsyncFunction,
+  type UseMutateFunction,
+} from "@tanstack/react-query";
+import type { Hex, TypedDataDefinition } from "viem";
+import { useAlchemyAccountContext } from "../context.js";
+import { ClientUndefinedError } from "../errors.js";
+import type { BaseHookMutationArgs } from "../types.js";
+import type { UseSmartAccountClientResult } from "./useSmartAccountClient.js";
+
+export type SignTypedDataArgs = { typedData: TypedDataDefinition };
+
+export type UseSignTypedDataMutationArgs = BaseHookMutationArgs<
+  Hex,
+  SignTypedDataArgs
+>;
+
+export type UseSignTypedDataArgs = {
+  client: UseSmartAccountClientResult["client"] | undefined;
+} & UseSignTypedDataMutationArgs;
+
+export type UseSignTypedDataResult = {
+  signTypedData: UseMutateFunction<Hex, Error, SignTypedDataArgs, unknown>;
+  signTypedDataAsync: UseMutateAsyncFunction<
+    Hex,
+    Error,
+    SignTypedDataArgs,
+    unknown
+  >;
+  signedTypedData: Hex | undefined;
+  isSigningTypedData: boolean;
+  error: Error | null;
+};
+
+export function useSignTypedData({
+  client,
+  ...mutationArgs
+}: UseSignTypedDataArgs): UseSignTypedDataResult {
+  const { queryClient } = useAlchemyAccountContext();
+
+  const {
+    mutate: signTypedData,
+    mutateAsync: signTypedDataAsync,
+    data: signedTypedData,
+    isPending: isSigningTypedData,
+    error,
+  } = useMutation(
+    {
+      mutationFn: async (params: SignTypedDataArgs) => {
+        if (!client) {
+          throw new ClientUndefinedError("useSignTypedData");
+        }
+        return client.signTypedDataWith6492({ ...params });
+      },
+      ...mutationArgs,
+    },
+    queryClient
+  );
+
+  return {
+    signTypedData,
+    signTypedDataAsync,
+    signedTypedData,
+    isSigningTypedData,
+    error,
+  };
```

Comment:
Quick thoughts: 
1. The `signedTypeData` and `error` imply that we hold some type of state. But we also provide the async version of the function. 
   1. This would be confusing for a downstream consumer of this hook. 
2. I know it seems inconvenient, but we should make it so that people explicitly use the onError or onSuccess callback. So it is more obvious how that would work then, with the signedTypeData and error results.
   1. We expose a separate hook called `useAsyncSignTypedData` that expects the user to use the async callback and handle the success and failures themselves.

What are your thoughts? That might also simplify the docs a bit. But realize this would be more code. Side note this has been a problem I've had with react-query for a long time 😬

Reply:
Doesn't the async method in react-query still plug in into the mutation lifecycle? So on success and on error still fire, as well as the loading states and error states? I kinda like the convenience of having the async method exported with the hooks, we should probs add it to some of the other hooks too actually because people might want to chain certain calls in their business logic, which is only doable with the async method. At which point having async versions of each hook becomes kind of cumbersome to maintain. 

Fwiw this is how wagmi handles it as well

Reply:
based on the useMutation docs, async does plug into the lifecycle: https://tanstack.com/query/latest/docs/framework/react/reference/useMutation#usemutation

I think it's a 2 way door to introduce a separate async hook later. agree it's cumbersome. esp. if wagmi handles things this way too, let's keep async mutate on the same hook as mutate.

Reply:
> Doesn't the async method in react-query still plug in into the mutation lifecycle? So on success and on error still fire, as well as the loading states and error states? I kinda like the convenience of having the async method exported with the hooks, we should probs add it to some of the other hooks too actually because people might want to chain certain calls in their business logic, which is only doable with the async method. At which point having async versions of each hook becomes kind of cumbersome to maintain.

It does. I'm not super opinionated about this, but this is a personal preference ergonomic issue that I think we'll only find out more about down the road. I agree on the less code we maintain the better. So let's punt on my idea, and keep it to one hook for now and then we can break it out later. Just might be a point of confusion later down the road.